File: 백준/Silver/16493. 최대 페이지 수/최대 페이지 수.js
const readline = require('readline').createInterface({
  input: process.stdin,
  output: process.stdout
});

let input = [];
let N, M, books;

readline.on('line', (line) => {
  input.push(line);
}).on('close', () => {
  [N, M] = input[0].split(' ').map(Number)
  books = input.slice(1).map(i => i.split(' ').map(Number));
  const result = solution();
  console.log(result);
  process.exit();
});

const solution = () => { 
  let DP = new Array(N+1).fill(0);
  books.sort((a, b) => b[0] - a[0]); 

  for (const [d, page] of books) {
    for (let i = N; i > 0; i--) {
      if (i-d >= 0) DP[i] = Math.max(DP[i], DP[i-d]+page); 
    }
  }

  return Math.max(...DP);
}

File: 백준/Silver/16493. 최대 페이지 수/최대 페이지 수.py
import sys  
input = sys.stdin.readline

N, M = map(int, input().split())
chapter = []
for _ in range(M):
  day, page = map(int, input().split())
  chapter.append([day, page])
 
dp = [[0 for _ in range(N+1)] for _ in range(M+1)]

for index, data in enumerate(chapter):
  day, page = data
  for i in range(1, N+1):
    if i-day >= 0:
      dp[index+1][i] = max(dp[index][i], dp[index][i-day]+page)
    else: 
      dp[index+1][i] = dp[index][i]
 
print(dp[M][N])